src/Target.cpp

```diff
@@ -753,6 +799,89 @@ int Target::natural_vector_size(const Halide::Type &t) const {
     }
 }
 
+bool Target::get_runtime_compatible_target(const Target& other, Target &result) {
+    // Create mask to select features that:
+    // (a) must be included if either target has the feature (union)
+    // (b) must be included if both targets have the feature (intersection)
+    // (c) must match across both targets; it is an error if one target has the feature and the other doesn't
+    const std::array<Feature, 15> union_features = {
+            // These are true union features.
+            CUDA, OpenCL, OpenGL, OpenGLCompute, Metal, D3D12Compute, NoNEON,
+
+            // These features are actually intersection-y, but because targets only record the _highest_,
+            // we have to put their union in the result and then take a lower bound.
+            CUDACapability30, CUDACapability32, CUDACapability35, CUDACapability50, CUDACapability61,
+            HVX_v62, HVX_v65, HVX_v66
+    };
+
+    const std::array<Feature, 12> intersection_features = {
+            SSE41, AVX, AVX2, FMA, FMA4, F16C, ARMv7s,VSX, AVX512, AVX512_KNL, AVX512_Skylake, AVX512_Cannonlake
+    };
+
+    const std::array<Feature, 10> matching_features = {
+            SoftFloatABI, Debug, TSAN, ASAN, MSAN, HVX_64, HVX_128, MinGW, HexagonDma, HVX_shared_object
+    };
+
+    // bitsets need to be the same width.
+    decltype(result.features) union_mask;
+    decltype(result.features) intersection_mask;
+    decltype(result.features) matching_mask;
+
+    for (auto& feature : union_features) {
+        union_mask.set(feature);
+    }
+
+    for (auto& feature : intersection_features) {
+        intersection_mask.set(feature);
+    }
+
+    for (auto& feature : matching_features) {
+        matching_mask.set(feature);
+    }
+
+    if (arch != other.arch || bits != other.bits || os != other.os) {
+        user_warning << "runtime targets must agree on platform (arch-bits-os)\n";
+        return false;
+    }
+
+    if ((features & matching_mask) != (other.features & matching_mask)) {
+        user_warning << "runtime targets must agree on SoftFloatABI, Debug, TSAN, ASAN, MSAN, HVX_64, HVX_128, MinGW, HexagonDma, and HVX_shared_object\n";
+        return false;
+    }
+
+    // Union of features is computed through bitwise-or, and masked away by the features we care about
+    // Intersection of features is computed through bitwise-and and masked away, too.
+    // We merge the bits via bitwise or.
+    Target output = Target{os, arch, bits};
+    output.features = ((features | other.features) & union_mask)
+            | ((features | other.features) & matching_mask)
+            | ((features & other.features) & intersection_mask);
+
+    // Pick tight lower bound for CUDA capability. Use fall-through to clear redundant features
+    switch (get_cuda_capability_lower_bound(output)) {
+        default: // no CUDA feature; clear all capability flags
+        case 20: output.features.reset(CUDACapability30);
```

Comment:
these should have comments about 'deliberately falling thru', since I presume that's what's happening here.

Reply:
Is that not what "Use fall-through to clear redundant features" is doing? Should I add comments after each line like:

    // Pick tight lower bound for CUDA capability. Use fall-through to clear redundant features
    switch (get_cuda_capability_lower_bound(output)) {
        default: // no CUDA feature; clear all capability flags
        case 20: output.features.reset(CUDACapability30); // fall-thru
        case 30: output.features.reset(CUDACapability32); // fall-thru
        case 32: output.features.reset(CUDACapability35); // fall-thru
        case 35: output.features.reset(CUDACapability50); // fall-thru
        case 50: output.features.reset(CUDACapability61); // fall-thru
        case 61: break;
    }

I personally think that's a little noisy given the comment right in front of the switch.

Reply:
I'd err on the side of noise; having fallthru be possible by accident is a misfeature of the language. (Google-internal code requires a specific annotation for deliberate fallthru; omitting it will trigger linter errors.)

Reply:
Okay, I can add those, then!